Jetty 12.0.x core security

gregw opened the pull request:
This is a core security mechanism, which will ultimately be the base of each of the EEn security handlers.

lachlan-roberts committed:
initial work on core security module

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-core-security

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12.0.x-core-security

gregw committed:
WIP

gregw committed:
BASIC auth working

gregw commented:
@janbartel @sbordet @lorban I've taken over @lachlan-roberts's core jetty-security work. 

This is still draft, but I would really appreciate some early feedback, specially on names and APIs.
  
The plan is in this PR to get a cleaned up mechanism working with all the common login & identity services, plus the authenticators and authentications.   This will be able to be used stand alone.
Only once we are happy with this mechanism as a stand along module, will I then update the eeN modules to actually use it.

Currently it is tested to BASIC auth and data constraints.   Best place to start looking is in `SecurityHandlerTest`


@janbartel yeah I know there is no javadoc. I will write it once the basic API is agreed apon.

gregw committed:
FORM auth working

gregw committed:
FORM auth working

gregw committed:
FORM auth POST working

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12.0.x-core-security

gregw committed:
WIP

Signed-off-by: gregw <gregw@webtide.com>

gregw committed:
WIP

Signed-off-by: gregw <gregw@webtide.com>

gregw committed:
WIP

gregw committed:
WIP

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-core-security

gregw committed:
WIP

Signed-off-by: gregw <gregw@webtide.com>

gregw committed:
WIP

gregw commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/authentication/DigestAuthenticator.java:
```diff
@@ -0,0 +1,376 @@
+
+            try
+            {
+                MessageDigest md = MessageDigest.getInstance("MD5");
```
spec requires MD5

lorban replied:
It may be worth a comment?

gregw commented:
Note I have a parallel branch jetty-12.0.x-core-security-ee9, where I am applying this to ee9, very much more a WIP... but it preserves the git history of the moved files.

gregw committed:
WIP auth parameters

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12.0.x-core-security

lorban commented on jetty-core/jetty-ee/src/main/java/org/eclipse/jetty/ee/security/ConstraintAware.java:
```diff
@@ -0,0 +1,68 @@
+    void addKnownRole(String role);
+
+    /**
+     * See Servlet Spec 31, sec 13.8.4, pg 145
```
Referring to the servlet spec in core sounds counter-productive.

gregw replied:
EE package is for common servlet stuff, so I think it is OK here.

lorban commented on jetty-core/jetty-ee/src/main/java/org/eclipse/jetty/ee/security/ConstraintAware.java:
```diff
@@ -0,0 +1,68 @@
+    boolean isDenyUncoveredHttpMethods();
+
+    /**
+     * See Servlet Spec 31, sec 13.8.4, pg 145
```
Referring to the servlet spec in core sounds counter-productive.

lorban commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Authenticator.java:
```diff
@@ -0,0 +1,137 @@
+     * manipulate the request to update it with information that
+     * can be inspected prior to validateRequest being called.
+     * The primary purpose of this method is to satisfy the Servlet
+     * Spec 3.1 section 13.6.3 on handling Form authentication
```
Another ref to the servlet spec.

lorban commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/SecurityHandler.java:
```diff
@@ -0,0 +1,761 @@
+    public static String SESSION_AUTHENTICATED_ATTRIBUTE = "org.eclipse.jetty.security.sessionAuthenticated";
+
+    private static final Logger LOG = LoggerFactory.getLogger(SecurityHandler.class);
+    private static final List<Authenticator.Factory> __knownAuthenticatorFactories = new ArrayList<>();
```
It feels like this field shouldn't be static but should be managed by `LifeCycle`'s `doStart`/`doStop`.

gregw replied:
But it is from the service loader, which is intrinsically static!

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/SecurityHandler.java:
```diff
@@ -0,0 +1,761 @@
+            return add(PathSpec.from(pathSpec), constraint);
+        }
+
+        public Constraint add(PathSpec pathSpec, Constraint constraint)
```
These methods should be renamed to `put` since they have put semantic (and the implementation calls a `put()` method).

sbordet commented on jetty-core/jetty-ee/src/main/java/org/eclipse/jetty/ee/security/ConstraintAware.java:
```diff
@@ -0,0 +1,68 @@
+     * @param constraintMappings the mappings
+     * @param roles the roles
+     */
+    void setConstraintMappings(List<ConstraintMapping> constraintMappings, Set<String> roles);
```
Apparently this method is not used.
I would remove it anyway, because if it should be called only during initialization, we can just pass the mappings and the roles to the constructor of the implementation, and avoid to have this setter in the interface.

sbordet commented on jetty-core/jetty-security/src/test/java/org/eclipse/jetty/security/SecurityHandlerTest.java:
```diff
@@ -0,0 +1,203 @@
+        assertThat(response, containsString(":9999"));
+        assertThat(response, not(containsString("You are OK")));
+
+        response = _connectorS.getResponse("GET /ctx/integral/info HTTP/1.0\r\nX-Forwarded-Proto: https\r\n\r\n");
```
Let's use everywhere the new `Forwarded` header, rather than the old `X-Forwarded`.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Constraint.java:
```diff
@@ -0,0 +1,211 @@
+ */
+public interface Constraint
+{
+    enum UserData
```
The enum name is hard to understand. I think this is usually referred to as `TransportGuarantee` which is longer but clearer (perhaps just `Transport` would do).

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Constraint.java:
```diff
@@ -0,0 +1,211 @@
+        }
+    }
+
+    enum Authorization
```
The enum name should be `Authentication`, not Authorization, because then its constants are all about `authenticated`.

gregw replied:
No this is authorization.  It is not about how/why we authenticate, it is about what authentication is required to pass a constraint.  So it is authorizing the authentication.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Constraint.java:
```diff
@@ -0,0 +1,211 @@
+     */
+    default String getName()
+    {
+        return "unnamed@%x".formatted(hashCode());
```
Feels like should be mandatory to implement? If so I'd make the method abstract.

gregw replied:
hmmm maybe.   I'm not 100% convinced of the fluent-like style I'm using here.   If a name was mandatory then every method would need to take the name of the new Constraint produced... unless we went for a full Builder pattern.

Currently with
```java
  Constraint constraint1 = Constraint.CONFIDENTIAL.with("admin","user").named("bob");
  Constraint constraint2 = Constraint.CONFIDENTIAL.named("tom").with("admin","user");
```
The two constraints are functionally the same, but 1 is named bob, whilst 2 is anonymous (it was based on tom, but a new constraint was made).

If name was mandatory, then we would need to do:
```java
  Constraint constraint1 = Constraint.CONFIDENTIAL.with("bob", "admin","user");
  Constraint constraint2 = Constraint.CONFIDENTIAL.with("tom","admin","user");
```
or go to a full builder pattern:

```java
  Constraint constraint1 = Constraint.CONFIDENTIAL.with("admin","user").named("bob").build();
  Constraint constraint2 = Constraint.CONFIDENTIAL.named("tom").with("admin","user").build();
```

I'm leaning towards full builder.  Why be different?

sbordet commented on jetty-core/jetty-security/src/main/java/module-info.java:
```diff
@@ -0,0 +1,24 @@
+
+module org.eclipse.jetty.security
+{
+    uses org.eclipse.jetty.security.Authenticator.Factory;
```
Put the `uses` as last, see why in #6938.

Also, can we hide in `.internal` many classes?
For example, `RolePrincipal`, `RoleRunAsToken`, etc.

gregw replied:
I think that those classes will more likely move to ee.  But if not, I'll see if they can be internal.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Authentication.java:
```diff
@@ -0,0 +1,228 @@
+     * <p>
+     * This convenience instance is for when an authentication challenge has been sent.
+     */
+    Authentication SEND_CONTINUE = new Challenge()
```
Should be called `CHALLENGE`.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Authentication.java:
```diff
@@ -0,0 +1,228 @@
+     * <p>
+     * This convenience instance is for when an authentication failure has been sent.
+     */
+    Authentication SEND_FAILURE = new Failure()
```
Drop `SEND`, just call it `FAILURE`?

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Authentication.java:
```diff
@@ -0,0 +1,228 @@
+            return "FAILURE";
+        }
+    };
+    Authentication SEND_SUCCESS = new SendSuccess()
```
Drop `SEND` just call it `SUCCESS`?

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Authenticator.java:
```diff
@@ -0,0 +1,137 @@
+    /**
+     * Authenticator Configuration
+     */
+    interface AuthConfiguration
```
Just call it `Configuration`, no need to repeat `Auth`.

gregw replied:
I agree it would be better, but there are naming clashes as `OpenIdConfiguration` is already used, so I can't rename `OpenIdAuthConfiguration`.   So will leave it for now.

sbordet replied:
But it is really confusing. It is "auth" as in "authentication" or "authorization"?
I don't understand the reason about `OpenIdConfiguration`, which is a different thing.
Just call it `OpenIdAuthenticatorConfiguration`.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Authenticator.java:
```diff
@@ -0,0 +1,137 @@
+     */
+    interface AuthConfiguration
+    {
+        String getAuthMethod();
```
I would expand to `getAuthenticationMethod()`.
Also, "Method" is overloaded, so perhaps `getAuthenticationMode()` is better?

gregw replied:
How about just `Authenticator.getName()`?

sbordet replied:
If this class get renamed to just `Configuration`, get it's not the name of the configuration, so I would like this to be more explicit even if longer, so `getAuthenticationMethod()` is my preference.

sbordet replied:
Bump.

sbordet replied:
@gregw this one has not been addressed.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Authenticator.java:
```diff
@@ -0,0 +1,137 @@
+     */
+    interface Factory
+    {
+        Authenticator getAuthenticator(Server server, Context context, AuthConfiguration configuration, IdentityService identityService, LoginService loginService);
```
Rename to `newAuthenticator()`.
Why does it need the last 2 parameters, since they can be extracted from the configuration?

gregw replied:
Fine question.  Don't know. Will remove and see what happens.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/SecurityHandler.java:
```diff
@@ -0,0 +1,761 @@
+     *
+     * @param loginService the loginService to set
+     */
+    public void setLoginService(LoginService loginService)
```
I find this very confusing (in previous Jetty versions), so I think we should improve in Jetty 12.

I would like to see just `SecurityHandler.setAuthenticator()`.

If the Authenticator needs a LoginService, we should have only `Authenticator.setLoginService`, not also a setter here in `SecurityHandler`.

If the `LoginService` needs an `IdentityService`, we should have only `LoginService.setIdentityService()`.

Better yet, it would be great if we can remove the setters from the interfaces altogether, so that you cannot create a `BasicAuthenticator` if you don't pass a `LoginService` to its constuctor; and you cannot create a `LoginService` if you don't pass an `IdentityService` to its constructor.

I say this because I've been bitten multiple times by setting up a `SecurityHandler` with an `Authenticator` only (and forgetting all the other pieces) only to get exceptions at runtime.

gregw replied:
yeah... however we sill still need some auto discovery mechanism for `loginService`.  The issue is that the Authenticator is selected and configured by the application, whilst the loginService is selected and configured by the deployer. 

So either we link them up in the SecurityHandler,  OR every app that configures Authentication needs to know how to discover the loginService.

The later is not necessarily bad, as we could do something like `LoginService.for(server, "realmName");`
 
will ponder....

gregw replied:
Cleaned up SPNEGO confusion

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/SecurityHandler.java:
```diff
@@ -0,0 +1,761 @@
+        throws Exception
+    {
+        // complicated resolution of login and identity service to handle
+        // many different ways these can be constructed and injected.
```
Yeah, too complicated.
Eventually it all boils down to 1 Authenticator, which needs 1 LoginService and 1 IdentityService. 
After doStart() is called, these don't change (they certainly cannot in the middle of handling a request, despite the setters), so I'd guess `ServletHandler -> Authenticator -> LoginService -> IdentityService` is enough?

No need for beans to lookup on the server, no need for setters repeated in multiple interfaces, no need for double check that all these objects are actually consistent otherwise throw ISE, etc.

gregw replied:
yeah this is all existing code.  I will look at how we actually do it in our modules/xml and simplify to support just that.

gregw replied:
As discussed, leaving this as is

sbordet commented on jetty-core/jetty-security/src/main/java/module-info.java:
```diff
@@ -0,0 +1,24 @@
+    requires java.security.jgss;
+    requires java.sql;
+
+    exports org.eclipse.jetty.security;
```
Package `org.eclipse.jetty.security.authentication` must be exported.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Authentication.java:
```diff
@@ -0,0 +1,228 @@
+    /**
+     * A successful Authentication with User information.
+     */
+    interface User extends LogoutAuthentication
```
Some of these nested interfaces are called `Authentication.LoginAuthentication` (so they repeat "Authentication"), some don't.
We should use one style only, I think the short one is my favorite, as long as it is used elsewhere as `Authentication.User` not just `User`.

gregw replied:
Agree... but i may leave changing that until I've refactored eeN to use these, as they are old names and once this is used, then refactoring will do the renames.

sbordet reviewed (changes requested):
@gregw perhaps we need to discuss my comments, I'm sure I'm missing important details.

gregw commented:
@sbordet your comments are mostly clear.

I agree on most of the renamings, however I might delay some until this is merged with eeX implementations, so the refactor will catch them..... or perhaps I won't and we can keep old names in the ee facades over the new names... except for ee10 which will expose the new names.    I'll ponder the timing.

Some of your comments are correctly pointing out complications inherited from the original implementation.   I'll work to simplify to just what is needed; move to a common ee; or just remove if they really are legacy.

I think I'll change the Constraint to a true Builder pattern as the fluent-like API almost works, but is surprising with regards to naming.

So thanks for the review, it's pointed out more areas that need to be de-legacied and simplified.   Lots to do on the plane trip :)

gregw committed:
WIP

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12.0.x-core-security

gregw committed:
WIP

gregw committed:
WIP

gregw committed:
WIP

gregw committed:
WIP

Signed-off-by: gregw <gregw@webtide.com>

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12.0.x-core-security

gregw committed:
WIP

gregw committed:
WIP on JAAS tests

gregw committed:
WIP on major simplification

gregw committed:
WIP on major simplification

lachlan-roberts committed:
Create OpenID module in jetty-core

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
implement OpenID module for jetty-core

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12.0.x-core-security

# Conflicts:
#	jetty-core/jetty-session/src/main/java/org/eclipse/jetty/session/SessionHandler.java
#	jetty-core/jetty-session/src/main/java/org/eclipse/jetty/session/SimpleSessionHandler.java
#	jetty-core/jetty-session/src/test/java/org/eclipse/jetty/session/SessionHandlerTest.java
#	jetty-core/jetty-session/src/test/java/org/eclipse/jetty/session/SimpleSessionHandler.java

lachlan-roberts committed:
remove isMandatory and use getConstraintAuthentication method for Authenticator

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

gregw committed:
WIP

gregw commented:
@lachlan-roberts @janbartel I think that the next steps are to use as much of core security as possible in ee8/9 and ee10.
In ee10, we should be able to use all of core-security and just layer on configuration.

For ee8/ee9  I don't think we can use the core Authenticators, but we should be able to use the core IdentityService and LoginService, as these things are set on the server and used by the context, so we need to be able to share an identity service and login service between ee8/9 and ee10 contexts.

@lachlan-roberts do you want to look at ee10 and I'll look at ee9.  @janbartel want to check both of us.

gregw committed:
WIP to using core for EE9

gregw committed:
WIP to using core for EE9

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-core-security

gregw committed:
WIP to using core for EE9

gregw committed:
WIP to using core for EE9

gregw committed:
WIP to using core for EE9

gregw committed:
Added core library to modules

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-core-security

gregw committed:
Added core library to modules

gregw committed:
fixed mod files

gregw committed:
fixed ee8 files

gregw committed:
fixed ee8 jaas

gregw committed:
static

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-core-security

gregw committed:
fixed jaas modules

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12.0.x-core-security

gregw committed:
Still fixing ee8 modules

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-core-security

gregw committed:
WIP

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-core-security

# Conflicts:
#	jetty-ee9/jetty-ee9-openid/src/test/java/org/eclipse/jetty/ee9/security/openid/OpenIdProvider.java
#	tests/test-distribution/test-ee9-distribution/src/test/java/org/eclipse/jetty/ee9/tests/distribution/openid/OpenIdProvider.java

lachlan-roberts committed:
fix merge of jetty-12.0.x openid changes

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

gregw committed:
Update openid to access session

gregw committed:
WIP

gregw committed:
Renamed "Core" environment to "Server"

Maintained a "Core" environment for use by deployer for apps that only use core classes plus whatever else is configured for the core environment.

gregw committed:
Merge branch 'jetty-12-server-environment' into jetty-12.0.x-core-security

gregw committed:
fixed jaas

gregw committed:
more jaas cleanup

gregw committed:
Server env exists only in startArgs.

gregw committed:
rename Server env to JVM

gregw committed:
Merge branch 'jetty-12-server-environment' into jetty-12.0.x-core-security

gregw committed:
rename of the rename of the rename

updates from review

gregw committed:
updates from review

gregw committed:
Merge branch 'jetty-12-server-environment' into jetty-12.0.x-core-security

gregw committed:
updates from review

janbartel committed:
Add new security jar to osgi container deployment

janbartel commented on jetty-core/jetty-session/src/main/java/org/eclipse/jetty/session/ManagedSession.java:
```diff
@@ -51,6 +51,7 @@ public class ManagedSession implements Session
     /**
      * Attribute set if the session is secure
      */
+    @Deprecated
```
This static is still used to indicate whether or not a session has been created _before_ a successful authentication.

gregw replied:
Yeah, that's why only deprecated.    Need to find a better way... or a better place for it.

janbartel commented on jetty-core/jetty-server/src/main/config/modules/openid.mod:
```diff
@@ -0,0 +1,10 @@
+# DO NOT EDIT - See: https://www.eclipse.org/jetty/documentation/current/startup-modules.html
```
Why is there an `openid.mod` file also in `jetty-openid` maven module?

sbordet replied:
This file should be removed and only keep the one in `jetty-openid`, right?

janbartel commented on jetty-core/jetty-server/src/main/config/modules/security.mod:
```diff
@@ -0,0 +1,10 @@
+# DO NOT EDIT - See: https://www.eclipse.org/jetty/documentation/current/startup-modules.html
+
```
I think both `security.mod` and `jaas.mod` should be in the `jetty-security` maven project.

gregw replied:
agree, but I've got to warm up my maven-foo to add a config component.   Will do, but will resolve other issues first....

janbartel commented on jetty-core/jetty-server/src/main/config/etc/jetty-jaas.xml:
Shouldn't this now be in the `jetty-security` maven module?

janbartel commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/UserIdentityScope.java:
```diff
@@ -0,0 +1,66 @@
+ * User object that encapsulates user identity and operations such as run-as-role actions,
+ * checking isUserInRole and getUserPrincipal.
+ * <p>
+ * Implementations of UserIdentity should be immutable so that they may be
```
```suggestion
 * Implementations of UserIdentityScope should be immutable so that they may be
```

Or are these 2 different classes altogether? Either way the javadoc needs fixing.

janbartel commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/SessionHandler.java:
```diff
@@ -693,6 +695,20 @@ public void onSessionPassivation(Session session)
 
     public class ServletSessionApi implements HttpSession, Session.API
     {
+        public static Function<Boolean, Session> newGetSession(ServletRequest servletRequest)
```
```suggestion
        public static Function<Boolean, Session> createOrGetSession(ServletRequest servletRequest)
```

sbordet replied:
I prefer `getOrCreateSession()`.

janbartel reviewed (changes requested):
Mostly comments, but I think you've got some duplicated files, so hence the -1.

gregw committed:
updates from review

gregw committed:
updates from review

gregw committed:
updates from review

gregw committed:
Merge branch 'jetty-12-server-environment' into jetty-12.0.x-core-security

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-core-security

gregw committed:
ee10 compiled with core security

gregw committed:
ee10 compiled with core security

gregw committed:
ee10 compiled with core security

gregw committed:
ee10 security tests

gregw committed:
WIP

gregw committed:
ee10 security tests

gregw committed:
ee10 jaspi

gregw committed:
ee10 jaspi

gregw committed:
ee10 jaspi

gregw committed:
fix javadoc

gregw committed:
WIP

gregw committed:
WIP

gregw committed:
WIP on ConstraintTest

gregw committed:
WIP on ConstraintTest

gregw committed:
WIP on ConstraintTest

gregw committed:
redirection tests

gregw committed:
more FormAuth fixes

gregw committed:
All ConstraintTests passing ee10

gregw committed:
All ConstraintTests passing ee10

gregw committed:
Fixed 302 to 303

gregw committed:
Refactor to remove forbidden as a separate boolean

gregw committed:
Bizarre servlet constraint combinations

gregw committed:
Fixed checkstyle

gregw committed:
Reverted 303 change

janbartel committed:
Add tests for combining of Constraints as per spec

gregw committed:
Reverted 303 change

gregw committed:
fixed jaas included

gregw committed:
get wrapped request from ServletContextResponse

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12.0.x-core-security

gregw committed:
get wrapped request from ServletContextResponse

gregw committed:
cherry picked new test from 10

gregw committed:
turned off debug

gregw committed:
fixed auth builder

gregw committed:
Some renaming after review

gregw commented:
Core security PR provides:

- An API and implementation in Core that can be used directly for sensible Authentication and Authorization
- EE8/EE9 uses parts of core Authentication (LoginService, IdentityService), but own Authenticators and Authorization Constraints
- EE10 builds on core authentication and authorisation.
- Lots of miscellaneous cleanups and fixes for servlet impl in both EE9 and EE10.

gregw committed:
Some renaming after review

gregw committed:
Some renaming after review

gregw committed:
Some renaming after review

gregw committed:
Really bad SPNEGOAuthenticator tests

sbordet committed:
Merged branch 'jetty-12.0.x' into 'jetty-12.0.x-core-security'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Restored HttpClientAuthenticationTest.
That required:
* Adding remove*() methods to RequestListeners (and associated tests).
* Avoiding null listeners in both RequestListeners and ResponseListeners.
* Nulling out the current chunk in HttpSender, as a read()+release()+read()+throw sequence could have double released the chunk read in the first read.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Restored SPNEGOAuthenticationTest.
That required to introduce SecurityUtils.doAs(),
to replace Subject.doAs() that is terminally deprecated.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed checkstyle.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed RequestListeners (doh!).
Fixed test that broke because of conversion of input string to text block.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw committed:
remove stack trace

gregw committed:
renames for clarity

gregw commented on jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/util/SPNEGOAuthenticationTest.java:
```diff
@@ -94,26 +124,21 @@ public void dispose() throws Exception
+        spnegoLoginService.setServiceName(serviceName);
+        spnegoLoginService.setHostName(serviceHost);
+        // Start the authorizationService along with the loginService.
+        spnegoLoginService.addBean(hashLoginService);
```
@sbordet Thanks for these updates... but there is something weird here.   To make the test work it needs a `HashLoginService` as a been and a `SPNEGOLoginService`, which wraps the HSL, set on the `Authenticator`.

It should just need the SLS, set either as a bean or directly on the `Authenticator`.

So what is looking for the HSL as a bean?

sbordet replied:
> So what is looking for the HSL as a bean?

Nothing is looking for it, as the comment says it's just for lifecycle purposes.

As for the need for both the `SPNEGOLoginService` and the `HashLoginService`, IIRC it's due to the fact that SPNEGO does authentication, but has no concept of roles.
So in order to retrieve the roles, another service is needed, and it just so happen that we can use any of the other existing that we already have.

So we can compose SPNEGO with Hash or JDBC, etc.

gregw replied:
I've no problem with the composition.  But if it wraps another login service, then it should be an auto-bean, so if it starts it then it manages it.

it is confusing to need one login service as a been and the other as a setter.   I'll fix.

sbordet committed:
Removed obsolete comment.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw committed:
SPNEGO cleanup

sbordet commented on documentation/jetty-documentation/src/main/asciidoc/operations-guide/jaas/chapter.adoc:
```diff
@@ -154,10 +154,10 @@ Refer to the link:https://docs.oracle.com/javase/7/docs/api/javax/security/auth/
+* link:{javadoc-url}/org.eclipse.jetty.security.jaas/spi/JDBCLoginModule.html[`org.eclipse.jetty.security.jaas.spi.JDBCLoginModule`]
+* link:{javadoc-url}/org.eclipse.jetty.security.jaas/spi/PropertyFileLoginModule.html[`org.eclipse.jetty.security.jaas.spi.PropertyFileLoginModule`]
+* link:{javadoc-url}/org.eclipse.jetty.security.jaas/spi/DataSourceLoginModule.html[`org.eclipse.jetty.security.jaas.spi.DataSourceLoginModule`]
+* link:{javadoc-url}/org.eclipse.jetty.security.jaas/spi/LdapLoginModule.html[`org.eclipse.jetty.security.jaas.ldap.LdapLoginModule`]
```
Restore paths to use a `/`, not a `.`.

lorban commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/RequestListeners.java:
```diff
@@ -35,25 +36,58 @@ public class RequestListeners implements Dumpable
-        addContentListener(listener);
-        addSuccessListener(listener);
-        addFailureListener(listener);
+        // Use logical OR to avoid short-circuit.
```
You mean _binary OR_ as the logical one is the one that does short circuit.

lorban replied:
Did you mean _binary OR_ as the logical one is the one that does short circuit?

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -266,8 +268,11 @@ default URI toURI()
+    class Immutable implements HttpURI, Serializable
     {
+        @Serial
+        private static final long serialVersionUID = 2245620284548399386L;
```
Why do we need this?

gregw replied:
I think this is needed for openid stuff that puts some things in the session.  @lachlan-roberts Do you recognise this?

sbordet commented on jetty-core/jetty-security/pom.xml:
```diff
@@ -0,0 +1,65 @@
+
+  <modelVersion>4.0.0</modelVersion>
+  <artifactId>jetty-security</artifactId>
+  <name>Core :: Security </name>
```
Remove space after `Security`.

sbordet commented on jetty-core/jetty-security/pom.xml:
```diff
@@ -0,0 +1,65 @@
+  <modelVersion>4.0.0</modelVersion>
+  <artifactId>jetty-security</artifactId>
+  <name>Core :: Security </name>
+  <description>The common jetty security implementation</description>
```
`jetty` -> `Jetty`.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Authenticator.java:
```diff
@@ -0,0 +1,213 @@
+ * Various types of {@link AuthenticationState} are returned in order to
+ * signal the next step in authentication.
+ *
+ * @version $Rev: 4793 $ $Date: 2009-03-19 00:00:01 +0100 (Thu, 19 Mar 2009) $
```
Remove this line.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/AuthenticationState.java:
```diff
@@ -0,0 +1,358 @@
+ */
+public interface AuthenticationState
+{
+    static AuthenticationState getAuthentication(Request request)
```
More javadocs.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Authenticator.java:
```diff
@@ -0,0 +1,213 @@
+        Authenticator getAuthenticator(Server server, Context context, AuthConfiguration configuration);
+    }
+
+    class Null implements Authenticator
```
Rename to `Empty` or `NoOp`?

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Constraint.java:
```diff
@@ -0,0 +1,329 @@
+ *  secure or not
+ *
+ */
+public interface Constraint
```
Javadocs.

lorban commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/RequestListeners.java:
```diff
@@ -35,25 +36,58 @@ public class RequestListeners implements Dumpable
-    public void addQueuedListener(Request.QueuedListener listener)
+    public boolean removeListener(Request.Listener listener)
     {
+        // Use logical OR to avoid short-circuit.
```
Did you mean _binary OR_ as the logical one is the one that does short circuit?

sbordet committed:
Updated comment about using the binary `|` operator.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/DefaultAuthenticatorFactory.java:
```diff
@@ -0,0 +1,97 @@
+        return authenticator;
+    }
+
+    /**
+     * @return the loginService
+     */
+    public LoginService getLoginService()
+    {
+        return _loginService;
+    }
+
+    /**
+     * @param loginService the loginService to set
+     */
+    public void setLoginService(LoginService loginService)
+    {
+        _loginService = loginService;
+    }
```
These methods are never used, and likewise the field.
What remains is the `if/else` in `getAuthenticator()`, which I think can be moved to the interface as `getDefaultAuthenticator()` perhaps, so we can get rid of this class?

gregw replied:
I've removed the unused methods, but I'll leave the class as is for now as it is used to plug in Jaspi.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/DefaultIdentityService.java:
```diff
@@ -0,0 +1,80 @@
+
+/**
+ *
+ */
```
Remove or fill in.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/DefaultIdentityService.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    private static final Association NOOP = () -> {};
+    private static final Association CLEAR_RUN_AS = () -> runAsRole.set(null);
```
Fields at the top of the class.

lorban commented on javadoc/pom.xml:
```diff
@@ -267,11 +267,11 @@
       <scope>provided</scope>
-    </dependency>-->
+    </dependency>
     <!--<dependency>
```
The `jetty-jaspi` artifact is gone, shouldn't this comment be removed?

BTW, `jetty-home-and-jetty-base.adoc` still refers to `jetty-jaas` and `jetty-jaspi`.

gregw replied:
yeah lots of work to do in the documentation

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/IdentityService.java:
```diff
@@ -24,43 +25,27 @@
+     * @return A {@link Closeable} that, when closed, will disassociate the token and restore any prior associations.
      */
-    Object setRunAs(UserIdentity user, RunAsToken token);
+    Association associate(UserIdentity user, RunAsToken runAsToken);
```
Having 2 methods is typically bad for implementers or overriders, because you override one, but forget the other.
Can we just have the method with 2 parameters and pass `null` if there is no `RunAsToken`?

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/SecurityHandler.java:
```diff
@@ -676,4 +634,74 @@ public String toString()
         }
     };
+
+    // TODO consider method mapping version
```
Remove this line, since it's done by `PathMapped` below?

gregw replied:
No, we don't yet support method matching.  So still is a TODO.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/SucceededAuthenticationState.java:
```diff
@@ -11,15 +11,14 @@
+package org.eclipse.jetty.security;
 
 /**
  * @version $Rev: 4793 $ $Date: 2009-03-19 00:00:01 +0100 (Thu, 19 Mar 2009) $
```
Remove this line.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/authentication/BasicAuthenticator.java:
```diff
@@ -76,29 +72,28 @@ public Authentication validateRequest(Request req, Response res, Callback callba
+    public static String authorization(String user, String password)
     {
-        return true;
+        return "basic " + Base64.getEncoder().encodeToString((user + ":" + password).getBytes(StandardCharsets.ISO_8859_1));
```
Per the specification, `basic` should be `Basic` with capital `B`.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/authentication/BasicAuthenticator.java:
```diff
@@ -76,29 +72,28 @@ public Authentication validateRequest(Request req, Response res, Callback callba
+        if (res.isCommitted())
+            return null;
 
         String value = "basic realm=\"" + _loginService.getName() + "\"";
```
`basic` -> `Basic`.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/internal/DefaultUserIdentity.java:
```diff
@@ -47,9 +50,8 @@ public Principal getUserPrincipal()
-        //Servlet Spec 3.1, pg 125
-        if ("*".equals(role))
-            return false;
+        if (DefaultIdentityService.isRoleAssociated(role))
+            return true;
```
This implementation is quite different from before.
Consider the before version had a spec reference, is the new one correct?

gregw replied:
The previous impl encoded servlet behaviours, which are now implemented in the layer above.  Specifically we convert a role of "*" into the `ANY_ROLE` authorization in ee10 servlet.  Also, only servlets do the role ref mapping.

I'm 98.5% sure that this impl is appropriate for core auth and can be used as a basis for servlet auth.

sbordet commented on jetty-core/jetty-security/src/test/java/org.eclipse.jetty.security.jaas/JAASLdapLoginServiceTest.java:
```diff
@@ -11,8 +11,9 @@
-package org.eclipse.jetty.ee10.jaas;
+package org.eclipse.jetty.security.jaas;
 
+/* TODO
```
Can this be restored?

gregw replied:
I might need help with this one.... @janbartel HELP!

sbordet commented on jetty-core/jetty-security/src/test/java/org.eclipse.jetty.security.jaas/JAASLoginServiceTest.java:
```diff
@@ -11,18 +11,9 @@
-import javax.security.auth.login.Configuration;
+package org.eclipse.jetty.security.jaas;
 
+/* TODO
```
Can this be restored?

sbordet commented on jetty-core/jetty-security/src/test/java/org.eclipse.jetty.security.jaas/spi/PropertyFileLoginModuleTest.java:
```diff
@@ -65,6 +62,7 @@ public void tearDown() throws Exception
     }
     
     @Test
+    @Disabled //TODO
```
Can this be restored?

gregw committed:
More javadoc

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-core-security

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/FormFields.java:
```diff
@@ -75,11 +75,26 @@ public static CompletableFuture<Fields> from(Request request, Charset charset)
         return from(request, charset, maxFields, maxLength);
     }
 
+    public static void set(Request request, CompletableFuture<Fields> fields)
+    {
+        request.setAttribute(FormFields.class.getName(), fields);
+    }
+
+    public static CompletableFuture<Fields> get(Request request)
+    {
+        Object attr = request.getAttribute(FormFields.class.getName());
+        if (attr instanceof FormFields futureFormFields)
+            return futureFormFields;
+        return EMPTY;
+    }
+
```
These methods are used only once, so I'd rather inline them, also because their signature and implementation is weird.
They should deal with `FormFields` but instead they deal with `CompletableFuture` (but not the getter?!?).
I think they don't belong here for just 1 usage.

gregw replied:
The reason they exist is that they hide the magic of how the form fields are associated with a request.  It is currently done via an attribute, but ultimately we might provide a field for it, or some other mechanism.
They deal with `CompletableFuture` because that's how FormFields are obtained, in that they may be read asynchronously.  It just so happens that servlets use a blocking get.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Response.java:
```diff
@@ -213,7 +218,32 @@ static <T extends Response.Wrapper> T as(Response response, Class<T> type)
-        sendRedirect(request, response, callback, HttpStatus.MOVED_TEMPORARILY_302, location, false);
+        int code = HttpMethod.GET.is(request.getMethod()) || request.getConnectionMetaData().getHttpVersion().getVersion() < HttpVersion.HTTP_1_1.getVersion()
+            ? HttpStatus.MOVED_TEMPORARILY_302
+            : HttpStatus.SEE_OTHER_303;
```
No need for this logic if you call the other method that has the same logic.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Session.java:
```diff
@@ -110,6 +110,8 @@ static Session getSession(Object session)
      */
     boolean isNew() throws IllegalStateException;
 
+    String encodeURL(Request request, String url, boolean cookiesInUse);
```
Javadocs.
Also, we don't use "URL" anymore, so rename to `encodeURI()` instead?

gregw committed:
Cleaned up delegate authentication for openid and SPNEGO

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Fields.java:
```diff
@@ -35,65 +35,59 @@
-            if (!value.equals(that.get(name), caseSensitive))
+            if (getSize() != that.getSize())
                 return false;
+            for (Map.Entry<String, Field> entry : fields.entrySet())
+            {
+                String name = entry.getKey();
+                Field value = entry.getValue();
+                if (!value.equals(that.get(name)))
+                    return false;
+            }
+            return true;
```
Isn't it easier to equals the maps, i.e. replacing with just `this.fields.equals(that.fields)`?

gregw replied:
perhaps.  But there is the case insensitivity of field names, but not values that comes into play.   So I'm happy with small changes to existing code rather than replacement

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IO.java:
```diff
@@ -465,6 +465,24 @@ public static void close(Closeable closeable)
         }
     }
 
+    /**
+     * Closes an arbitrary closable, and logs exceptions at ignore level
+     *
+     * @param closeable the closeable to close
+     */
+    public static void close(AutoCloseable closeable)
+    {
+        try
+        {
+            if (closeable != null)
+                closeable.close();
+        }
+        catch (Exception ignore)
+        {
+            LOG.trace("IGNORED", ignore);
+        }
+    }
+
```
There is already a `close(Closeable)`, I don't think this is needed, as I would fold both into this one (we are never interested in exceptions thrown from `close()`).

gregw replied:
But `Closeable`  extends `AutoCloseable`, so it is possible to just have an `AutoCloseable` that is not `Closeable`!

sbordet replied:
Ok, so remove `close(Closeable)`, not need to have 2 methods that are almost identical!

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/SecurityHandler.java:
```diff
@@ -46,28 +49,32 @@
+ * TODO rename to ConstraintHandler
  */
-public abstract class SecurityHandler extends Handler.Wrapper implements Authenticator.AuthConfiguration
+public abstract class SecurityHandler extends Handler.Wrapper implements AuthConfiguration
```
Rename to `ConstraintHandler`?

gregw replied:
yeah... I'm on the fence.  It is a better name, but it will change a lot of legacy examples, doco etc.
Let's fix all the other issues and ponder this one.

sbordet commented on jetty-ee10/jetty-ee10-bom/pom.xml:
```diff
@@ -61,8 +61,8 @@
-        <groupId>org.eclipse.jetty.ee10</groupId>
-        <artifactId>jetty-ee10-jaas</artifactId>
+        <groupId>org.eclipse.jetty</groupId>
+        <artifactId>jetty-security</artifactId>
```
I don't think this should be here.
Here we only want ee10 dependencies, and `jetty-security` is a transitive of `jetty-ee10-servlet`.

gregw replied:
I thought bom poms are meant to be exhaustive?    Eitherway, @olamy can you have a look at all the boms in this PR, as I have probably screwed them one way or another with structure changes.

sbordet commented on jetty-ee10/jetty-ee10-bom/pom.xml:
```diff
@@ -76,8 +76,8 @@
-        <groupId>org.eclipse.jetty.ee10</groupId>
-        <artifactId>jetty-ee10-openid</artifactId>
+        <groupId>org.eclipse.jetty</groupId>
+        <artifactId>jetty-openid</artifactId>
```
We don't want this dependency here, as this is ee10.

gregw replied:
But ee10 uses core openid.  @olamy help?

sbordet commented on jetty-ee10/jetty-ee10-demos/jetty-ee10-demo-embedded/src/main/java/org/eclipse/jetty/ee10/demos/LikeJettyXml.java:
```diff
@@ -217,7 +217,7 @@ public static Server createServer(int port, int securePort, boolean addDebugList
             throw new FileNotFoundException("Unable to find config: " + realmFile);
         login.setConfig(realmResource);
-        login.setHotReload(false);
+        login.setRefreshInterval(0);
```
Would rename this method to `setReloadInterval(long)`, as "refresh" does not really convey a special meaning.

sbordet commented on jetty-ee10/jetty-ee10-jaspi/src/main/java/org/eclipse/jetty/ee10/security/jaspi/JaspiAuthenticator.java:
```diff
@@ -255,17 +247,17 @@ public Authentication validateRequest(JaspiMessageInfo messageInfo) throws Serve
 
-    @Override
-    public boolean secureResponse(Request request, Response response, Callback callback, boolean mandatory, User validatedUser) throws ServerAuthException
+    // TODO
```
Is there something to do here? If so perhaps add what exactly?

sbordet replied:
Bump.

sbordet commented on jetty-ee10/jetty-ee10-jaspi/src/main/java/org/eclipse/jetty/ee10/security/jaspi/JaspiMessageInfo.java:
```diff
@@ -106,42 +105,31 @@ public String getAuthMethod()
-        return _map.isAuthMandatory();
-    }
-
     //TODO this has bugs in the view implementations.  Changing them will not affect the hardcoded values.
```
Should this TODO be done?

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java:
```diff
@@ -3032,7 +3015,7 @@ public String getContextPath()
         public String toString()
         {
-            return "ServletContext@" + ServletContextHandler.this.toString();
+            return "ServletContext@" + ServletContextHandler.this;
```
I would use `getClass().getSimpleName()` as everywhere else, because it will resist renamings (this class is not called `ServletContext`).

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextResponse.java:
```diff
@@ -89,6 +90,12 @@ public ServletContextResponse(ServletChannel servletChannel, ServletContextReque
+    public Request getRequest()
+    {
+        return _request;
+    }
```
```suggestion
    @Override
    public ServletContextRequest getRequest()
    {
        return (ServletContextRequest)super.getRequest();
    }
```

Also both the `_request` and `_response` fields are duplicates of what exists already in base classes, so there is no need to duplicate them (saves runtime memory).

sbordet commented on jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/JndiConfiguration.java:
```diff
@@ -20,7 +20,7 @@
  * <p>JNDI Configuration</p>
  * <p>This configuration configures the WebAppContext system/server classes to
- * be able to see the org.eclipse.jetty.jaas package.
+ * be able to see the org.eclipse.jetty.security.jaas package.
```
Fix javadocs copy/paste error, this class exposes `*.jndi.*`.

sbordet commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/UserIdentityScope.java:
```diff
@@ -0,0 +1,66 @@
+        }
+        return role;
+    }
+
```
Empty line!!!

sbordet commented on jetty-ee9/jetty-ee9-openid/src/main/java/module-info.java:
```diff
@@ -14,10 +14,11 @@
     requires org.eclipse.jetty.util.ajax;
-
+    requires transitive org.eclipse.jetty.security;
+    requires transitive org.eclipse.jetty.security.openid;
```
Please move these to alphabetical order.

sbordet commented on jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/JndiConfiguration.java:
```diff
@@ -20,7 +20,7 @@
  * <p>JNDI Configuration</p>
  * <p>This configuration configures the WebAppContext system/server classes to
- * be able to see the org.eclipse.jetty.jaas package.
+ * be able to see the org.eclipse.jetty.security.jaas package.
```
Fix copy/paste typo, this class exports `jndi`.

lorban commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/IdentityService.java:
```diff
@@ -24,43 +25,27 @@
-     * @param token RUNAS returned from previous associateRunAs call
-     */
-    void unsetRunAs(Object token);
+    void onLogout(UserIdentity user);
```
Javadoc?

gregw committed:
Updates from review

+ some renaming
+ more javadoc
+ Modified Constraint so that it can have "don't care" null values and combination is now order specific from least specific to most specific.

gregw committed:
Updates from review

+ some renaming
+ more javadoc
+ Modified Constraint so that it can have "don't care" null values and combination is now order specific from least specific to most specific.

gregw committed:
Updates from review

+ some renaming
+ more javadoc

gregw committed:
Updates from review

gregw commented:
@sbordet @lorban @janbartel thanks for the reviews.  I've made a first pass at fixing the issues you raised.
However, I noticed that the way I was combine core Constraints was just wrong.    So I have updated that a fair bit (as well as javadoc'd it).   I could describe the change here, but maybe it would be a good test of the javadoc if you can work out what it does now?!?!?

There are a couple of suggested renames that I think are good... but I'm still cautious about them.

gregw committed:
Fixes from updates from review

gregw committed:
Fixes from updates from review

gregw committed:
Fixes from updates from review

gregw committed:
Fixes from updates from review

gregw committed:
Fixes from updates from review

gregw committed:
Use INHERIT instead of null for "don't care"

gregw committed:
Fixes from updates from review

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12.0.x-core-security

gregw committed:
resovled some TODOs

gregw committed:
resolved some TODOs

gregw committed:
Moved some classes to better locations

sbordet commented on jetty-core/jetty-security/src/main/java/module-info.java:
```diff
@@ -11,17 +11,17 @@
+    requires transitive org.eclipse.jetty.util;
+    requires transitive org.slf4j;
+    requires java.security.jgss;
+    requires java.sql;
```
I understand these 2 are "optional" in Maven terms, because if you don't use SPNEGO you don't need `java.security.jgss`, and same for `java.sql`.

So technically they should be `requires static`. Although I am not sure what happens if you make them so 😁

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/AbstractLoginService.java:
```diff
@@ -84,7 +82,7 @@ public String toString()
 
     @Override
-    public UserIdentity login(String username, Object credentials, ServletRequest request)
+    public UserIdentity login(String username, Object credentials, Request request, Function<Boolean, Session> getSession)
```
Rename last argument to `getOrCreateSession`?

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/AuthenticationState.java:
```diff
@@ -0,0 +1,348 @@
+        /**
+         * @return The method used to authenticate the user.
+         */
+        String getAuthMethod();
```
"Auth" is really confusing. Can it be expanded to either `Authentication` or `Authorization`?

gregw replied:
I've got for `getAuthenticatorName` and changed the method on `Authenticator` to be just `getName`

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Authenticator.java:
```diff
@@ -0,0 +1,212 @@
+    /**
+     * @return The name of the authentication method
+     */
+    String getAuthMethod();
```
"Auth" -> either `Authentication` or `Authorization`.

gregw replied:
I've gone with `getName`

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Constraint.java:
```diff
@@ -0,0 +1,380 @@
+ */
+public interface Constraint
+{
+    Logger LOG = LoggerFactory.getLogger(Constraint.class);
```
We do not want this to be `public`, so please remove it, plus it's only used once.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Constraint.java:
```diff
@@ -0,0 +1,380 @@
+         * Access allowed only for authenticated user with specific role(s).
+         */
+        SPECIFIC_ROLE,
+
```
Remove empty line.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Constraint.java:
```diff
@@ -0,0 +1,380 @@
+     */
+    Set<String> getRoles();
+
+    default Builder builder()
```
Should be `static` so `Constraint.builder()` can be called directly.

gregw replied:
It was intended to use the `this` as the template for the builder, so can't be static.
But nobody uses it anyway, so I just deleted!

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Constraint.java:
```diff
@@ -0,0 +1,380 @@
+        private Set<String> _roles;
+        private Transport _transport;
+
+        public Builder()
```
Since there is `Constraint.builder()`, this should be `private`.

gregw replied:
That method was for something different (see above).

I'm OK with normal constructors

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Constraint.java:
```diff
@@ -0,0 +1,380 @@
+        public Builder()
+        {}
+
+        public Builder(Constraint constraint)
```
Make it `private`, adding `constraint(Constraint)` to the builder.

gregw replied:
ditto

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Constraint.java:
```diff
@@ -0,0 +1,380 @@
+     */
+    static Constraint combine(Constraint leastSpecific, Constraint mostSpecific)
+    {
+        String name = LOG.isDebugEnabled() ? leastSpecific.getName() + ">" + mostSpecific : null;
```
Always give a name and get rid of the `LOG`.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Constraint.java:
```diff
@@ -0,0 +1,380 @@
+        if (!roleSet.isEmpty() && auth != Authorization.SPECIFIC_ROLE)
+            throw new IllegalStateException("Constraint with roles must be SPECIFIC_ROLE, not " + auth);
+
+        return new Constraint()
```
Can we not use the builder here?

gregw replied:
The builder is implemented to use from, so it ends up here and uses this inner class.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/HashLoginService.java:
```diff
@@ -36,7 +36,7 @@ public class HashLoginService extends AbstractLoginService
 
     private Resource _config;
-    private int refreshInterval; // default is not to reload
+    private int reloadInterval; // default is not to reload
```
Make it `long` like we always do for time intervals.

gregw replied:
The units are seconds.  We use long for milliseconds.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/IdentityService.java:
```diff
@@ -76,10 +52,20 @@
 
     UserIdentity getSystemUserIdentity();
+
+    interface Association extends AutoCloseable
```
Javadocs.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/PropertyUserStore.java:
```diff
@@ -52,7 +52,7 @@ public class PropertyUserStore extends UserStore implements Scanner.DiscreteList
     protected Resource _configResource;
     protected Scanner _scanner;
-    protected int _refreshInterval = 0;
+    protected int _reloadInterval = 0;
```
Make it `long`.

gregw replied:
nah, it is in seconds.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/SPNEGOLoginService.java:
```diff
@@ -322,4 +320,29 @@ private GSSContextHolder(GSSContext gssContext)
+     * <p>A service to query for user roles.</p>
+     */
+    @FunctionalInterface
+    public interface AuthorizationService
```
This is not used anymore, delete?

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/SecurityHandler.java:
```diff
@@ -0,0 +1,769 @@
+    private LoginService _loginService;
+    private IdentityService _identityService;
+    private boolean _renewSession = true;
+    AuthenticationState.Deferred _deferred;
```
Make it `private` like the others.

sbordet commented on jetty-core/jetty-session/src/main/java/org/eclipse/jetty/session/SessionManager.java:
```diff
@@ -80,43 +80,45 @@ public interface SessionManager extends LifeCycle, SessionConfig
+
     int getSessionsCreated();
 
+    String encodeURL(Request request, String url, boolean cookiesInUse);
```
Rename to `encodeURI` for consistency.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IO.java:
```diff
@@ -459,9 +407,27 @@ public static void close(Closeable closeable)
+     *
+     * @param closeable the closeable to close
+     */
+    public static void close(AutoCloseable closeable)
```
Remove `close(Closeable)`, which is almost identical to this one.

gregw committed:
Renamed Authenticator.Configuration

gregw committed:
updates from a review

gregw committed:
updates from a review

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Constraint.java:
```diff
@@ -0,0 +1,362 @@
+            @Override
+            public String getName()
+            {
+                return name == null ? "unnamed@%x".formatted(hashCode()) : name;
```
Given that roles and authorization are normalized from `null` to a non-null value in `from()`, can we do the same for name and transport? Seems weird that we initialize some, but then have ternary expressions in the getters for others.
Doing so will also normalize `toString()` which now would print a `null` transport instead of `INHERIT`.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Constraint.java:
```diff
@@ -0,0 +1,362 @@
+            @Override
+            public Transport getTransport()
+            {
+                return transport == null ? Transport.INHERIT : transport;
```
Ditto.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/IdentityService.java:
```diff
@@ -76,10 +52,27 @@
     UserIdentity getSystemUserIdentity();
+
+    /**
+     * An associate between an identity and the current thread that can be terminated by closing.
```
```suggestion
     * An association between an identity and the current thread that can be terminated by {@link #close()}.
```

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/SecurityHandler.java:
```diff
@@ -0,0 +1,769 @@
+        {
+            return switch (group)
+            {
+                case EXACT -> 5;
+                case ROOT -> 4;
+                case SUFFIX_GLOB -> 3;
+                case MIDDLE_GLOB -> 2;
+                case PREFIX_GLOB -> 1;
+                case DEFAULT -> 0;
```
Should not `PathSpecGroup` implement `Comparable` instead?
You would do `g1.compareTo(g2)` where necessary.

gregw replied:
It does, but it implements the servlet spec order.  Besides, doing it this way makes it extensible in specialized securityhandlers

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IO.java:
```diff
@@ -452,27 +400,37 @@ public static boolean isEmptyDir(File dir)
+     *
+     * @param closeable the closeable to close
+     */
+    public static void close(Closeable closeable)
+    {
+        close((AutoCloseable)closeable);
+    }
```
Cannot be removed?

gregw replied:
It should be able to be, but broke CI big time.
Jenkins is a mystery to me, so simplest to just leave it in:)

sbordet reviewed (changes requested):
A few more nits/renaming, almost there.

gregw committed:
Update jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/IdentityService.java

Co-authored-by: Simone Bordet <simone.bordet@gmail.com>

gregw committed:
comment usage of MD5

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12.0.x-core-security

gregw committed:
removed redundant request field

gregw committed:
removed redundant response field

gregw commented:
@sbordet thanks for all the reviews.  I have fixed the last round and gone through and resolved the conversations that I think are fixed.   There are still a few left (boms, TODOs, etc.) but I think most of those are problems already in HEAD, so I would like to go for a merge and create new issues to fix the TODOs, boms etc.   So have a look at the remaining unresolved conversations and see which ones you'd like fixed before merge.

sbordet reviewed (changes requested):
@gregw I need one last rename, and then I'm good: `Authorization.Configuration.getAuthMethod()` -> `getAuthenticationName()` or similar, as long as it does not contain abbreviated "auth".

Also, there are now CI failures?

gregw committed:
updates from a review

gregw committed:
updates from a review

sbordet commented on jetty-core/jetty-openid/src/main/config/etc/jetty-openid.xml:
```diff
@@ -27,13 +27,13 @@
         <Arg name="clientId"><Property name="jetty.openid.clientId"/></Arg>
         <Arg name="clientSecret"><Property name="jetty.openid.clientSecret"/></Arg>
-        <Arg name="authMethod"><Property name="jetty.openid.authMethod" default="client_secret_post"/></Arg>
+        <Arg name="authenticatorName"><Property name="jetty.openid.authenticatorName" deprecated="jetty.openid.authMethod" default="client_secret_post"/></Arg>
```
I don't think this is right.
Elsewhere the `authenticatorName` is `OPENID`, while here is `client_secret_post`.
This should really be the client-side mechanism to use, so it should have a different name.
Perhaps `authenticationMethod` or `clientAuthentication[Method|Mode]`?

gregw replied:
@lachlan-roberts HELP!!! what is "client_secret_post" about?  Why can the openid authenticator change it's method name?

@sbordet I agree it looks strange.... but it started strange so maybe I open and issue, assign to Lachlan and we move on?

sbordet commented on jetty-core/jetty-openid/src/main/config/modules/openid.mod:
```diff
@@ -46,8 +43,8 @@ etc/jetty-ee10-openid.xml
 
-## What authentication method to use with the Token Endpoint (client_secret_post, client_secret_basic).
-# jetty.openid.authMethod=client_secret_post
+## What authenticator to use with the Token Endpoint (client_secret_post, client_secret_basic).
+# jetty.openid.authenticatorName=client_secret_post
```
Ditto.

gregw replied:
@lachlan-roberts ditto

sbordet commented on jetty-core/jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdConfiguration.java:
```diff
@@ -50,7 +50,7 @@ public class OpenIdConfiguration extends ContainerLifeCycle
     private final String clientSecret;
     private final List<String> scopes = new ArrayList<>();
-    private final String authMethod;
+    private final String authenticatorName;
```
Ditto.
This file has numerous occurrences of this to fix.

gregw replied:
@lachlan-roberts ditto ditto ditto ditto

sbordet commented on jetty-core/jetty-openid/src/test/java/org/eclipse/jetty/security/openid/OpenIdReamNameTest.java:
```diff
@@ -11,15 +11,13 @@
 //
 
-package org.eclipse.jetty.ee9.security.openid;
+package org.eclipse.jetty.security.openid;
```
Please rename this file to `OpenIdRealmNameTest` (not the missing `l` in `Ream`).

gregw replied:
glad you pointed out the missing l.  I read it 10 times and thought it was the same :)

sbordet commented on jetty-ee10/jetty-ee10-jaspi/src/main/java/org/eclipse/jetty/ee10/security/jaspi/JaspiMessageInfo.java:
```diff
@@ -101,69 +100,54 @@ public void setResponseMessage(Object response)
-    public boolean isAuthMandatory()
-    {
-        return _map.isAuthMandatory();
+        return _map.getAuthenticatorName();
```
This method is never used.
`_map.getAuthenticatorName()` is only ever used here.
I think we can totally remove the field (and the logic) from `MIMap`, and that should also allow to remove the delegation by removing `MIMap` entirely, which in turn fixes the TODO present in the `MIMap` class (and hopefully also fixes the use of generics for the Map?).

gregw replied:
I've removed the unused method, but for any other jaspi cleanup, I'd prefer to punt that to a new issue/PR.    I think there is a lot that needs careful thought about JASPI, so I'd prefer not to delay this PR whilst I page in what the hell JASPI is again and do a clean up.

gregw committed:
updates from review

gregw committed:
Merge branch 'jetty-12.0.x' into jetty-12.0.x-core-security

sbordet commented:
@gregw just a reminder:

* `Authenticator.getName() -> getAuthenticationType()`
* `OpenIdConfiguration.getAuthenticatorName() -> getAuthenticationMethod()`

Almost there!

gregw committed:
updates from review

gregw committed:
finish renaming
